src/btree/bt_handle.c

```diff
@@ -524,9 +524,26 @@ __btree_conf(WT_SESSION_IMPL *session, WT_CKPT *ckpt)
 
     /* If this is the first time opening the tree this run. */
     if (F_ISSET(session, WT_SESSION_IMPORT) || ckpt->run_write_gen < conn->base_write_gen)
-        btree->base_write_gen = btree->run_write_gen = btree->write_gen;
+        btree->run_write_gen = btree->write_gen;
     else
-        btree->base_write_gen = btree->run_write_gen = ckpt->run_write_gen;
+        btree->run_write_gen = ckpt->run_write_gen;
+
+    /*
+     * Don't reset the transaction ids in recovery because rollback to stable in addition to stable
+     * timestamp also depends on transaction ids from the page that are read into cache to decide if
+     * an update needs to be rolled back.
+     *
+     * To achieve that, use the last checkpointed run write generation number as base write
+     * generation number. The transaction ids are retained only on the pages that are written after
+     * the restart.
+     *
+     * Rollback to stable does not operate on logged tables and metadata, so it is skipped.
+     */
+    if (!F_ISSET(conn, WT_CONN_RECOVERING) || WT_IS_METADATA(btree->dhandle) ||
+      __wt_btree_immediately_durable(session))
+        btree->base_write_gen = btree->run_write_gen;
+    else
+        btree->base_write_gen = ckpt->run_write_gen;
```

Comment:
Is the idea that you want the `btree->base_write_gen` to be temporarily lower than any of the cells we're reading in via rollback to stable? Is this effectively the same as just setting it to 0?

Reply:
No. The earlier problem of RTS was reading the pages that are modified before the last restart.
Eg steps:
1. A page was written with txn id 100
2. restart -1
3. some more operations but the original page was not touched.
4. restart -2 --> This restart is reading the page with txnid 100 and remove its updates as part of RTS.

if we use 0, we will get the same problem, so we want to set the base write generation number as the run write generation number at restart -1, so any page that is modified after restart-1 will have transaction ids and rest of the pages will get reset.

Reply:
That's interesting. I think I get it.